lambda/debuginfo.ml

```diff
@@ -20,12 +20,30 @@ open Location
 module Scoped_location = struct
   type scope_item =
     | Sc_anonymous_function
-    | Sc_value_definition of string
-    | Sc_module_definition of string
-    | Sc_class_definition of string
-    | Sc_method_definition of string
+    | Sc_value_definition
+    | Sc_module_definition
+    | Sc_class_definition
+    | Sc_method_definition
 
-  type scopes = scope_item list
+  type scopes =
+    | Empty
+    | Cons of {item: scope_item; str: string; mutable str_fun: string}
+
+  module WeakString = Weak.Make(struct include String let hash = Hashtbl.hash end)
+
+  let memo_str = WeakString.merge (WeakString.create 128)
```

Comment:
I would be in favor of avoiding weak pointers to simplify the runtime surface fo the code, especially given that the sharing gains reported are rather small.

Reply:
As discussed during today's meeting, I've reverted this hash-consing.

Note that if the concern is just the use of Weak tables, one could just use a regular table, and clear it after each compilation unit / toplevel phrase.  But it was agreed that the tiny lack of sharing is not worth the trouble.